templates/components/multiagent/typescript/workflow/single-agent.ts

```diff
@@ -0,0 +1,236 @@
+      this.memory.put({ role: "system", content: this.systemPrompt });
+    }
+    this.memory.put({ role: "user", content: message });
+    return new InputEvent({ input: this.chatHistory });
+  }
+
+  private async handleLLMInput(
+    ctx: Context,
+    ev: InputEvent,
+  ): Promise<StopEvent<string | AsyncGenerator> | ToolCallEvent> {
+    if (ctx.get("streaming")) {
+      return await this.handleLLMInputStream(ctx, ev);
+    }
+
+    const result = await this.llm.chat({
+      messages: this.chatHistory,
+      tools: this.tools,
+    });
+    this.memory.put(result.message);
+
+    const toolCalls = this.getToolCallsFromResponse(result);
+    if (toolCalls.length) {
+      return new ToolCallEvent({ toolCalls });
+    }
+    this.writeEvent("Finished task", ctx);
+    return new StopEvent({ result: result.message.content.toString() });
+  }
+
+  private async handleLLMInputStream(
+    context: Context,
+    ev: InputEvent,
+  ): Promise<StopEvent<AsyncGenerator> | ToolCallEvent> {
+    const { llm, tools, memory } = this;
+    const llmArgs = { messages: this.chatHistory, tools };
+
+    const responseGenerator = async function* () {
+      const responseStream = await llm.chat({ ...llmArgs, stream: true });
+
+      let fullResponse = null;
+      let yieldedIndicator = false;
+      for await (const chunk of responseStream) {
+        const hasToolCalls = chunk.options && "toolCall" in chunk.options;
+        if (!hasToolCalls) {
+          if (!yieldedIndicator) {
+            yield false;
+            yieldedIndicator = true;
+          }
+          yield chunk;
+        } else if (!yieldedIndicator) {
+          yield true;
+          yieldedIndicator = true;
+        }
+
+        fullResponse = chunk;
+      }
+
+      if (fullResponse) {
+        memory.put({
+          role: "assistant",
+          content: "",
+          options: fullResponse.options,
+        });
+        yield fullResponse;
+      }
+    };
+
+    const generator = responseGenerator();
+    const isToolCall = await generator.next();
+    if (isToolCall.value) {
+      const fullResponse = await generator.next();
+      const toolCalls = this.getToolCallsFromResponse(
+        fullResponse.value as ChatResponseChunk<ToolCallLLMMessageOptions>,
+      );
+      return new ToolCallEvent({ toolCalls });
+    }
+
+    this.writeEvent("Finished task", context);
+    return new StopEvent({ result: generator });
+  }
+
+  private async handleToolCalls(
+    ctx: Context,
+    ev: ToolCallEvent,
+  ): Promise<InputEvent> {
+    const { toolCalls } = ev.data;
+
+    const toolMsgs: ChatMessage[] = [];
+
+    for (const call of toolCalls) {
+      const targetTool = this.tools.find(
+        (tool) => tool.metadata.name === call.name,
+      );
+      // TODO: make logger optional in callTool in framework
+      const toolOutput = await callTool(targetTool, call, {
+        log: () => {},
+        error: console.error.bind(console),
+        warn: () => {},
+      });
```

Comment:
@himself65 using `callTool` from the framework is very helpful, but would be nice if the logger object would be optional